Improve scan speed by avoiding division and modulus

fickludd opened the pull request:
UPDATED: reworked the implementation after some feedback. This removed some of the `PageCursor.setOffset` calls and further improved speed, now I get the following benchmark results locally:

```
                          3.4                            this PR
| --------------------- | --------------------------- | ---------------------------- | 
| all-node scan         | 328446.771 ± 2183.832 us/op | 200417.410 ± 1088.529  us/op |
| all-relationship scan | 428465.630 ± 2901.997 us/op | 325505.176 ± 50896.297 us/op |
```

ragadeeshu commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/DefaultNodeCursor.java:
```diff
@@ -85,6 +87,7 @@ void single( long reference, Read read )
         this.next = reference;
         //This marks the cursor as a "single cursor"
         this.highMark = NO_ID;
+        this.lastStoreReference = -2;
```
Why -2? I don't see it, could perhaps be clearer.

ragadeeshu reviewed (commented):
Just one tiny little comment.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/DefaultNodeCursor.java:
```diff
@@ -195,8 +198,15 @@ else if ( hasChanges && txs.nodeIsDeletedInThisTx( next ) )
             else
             {
+                lastStoreReference = next;
                 read.node( this, next++, pageCursor );
```
Would it be possible to assign `lastStoreReference` after the `next++` lines, initialise `lastStoreReference` to `-2`, aka. `NO_ID`, and change `else if ( lastStoreReference + 1 == next )` to `else if ( lastStoreReference == next )` ? Maybe then `lastStoreReference` could be called `nextScanReference`. Perhaps that would make the code clearer.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -1051,22 +1050,54 @@ void readIntoRecord( long id, RECORD record, RecordLoad mode, PageCursor cursor
+    private void readRecordFromPage( long id, RECORD record, RecordLoad mode, PageCursor cursor, int offset )
+            throws IOException
+    {
+        // There is a page in the store that covers this record, go read it
```
I think this comment can be removed now.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -1051,22 +1050,54 @@ void readIntoRecord( long id, RECORD record, RecordLoad mode, PageCursor cursor
+            if ( offset >= storeFile.pageSize() || pageId < 0 )
+            {
+                offset = 0;
+                if ( !cursor.next( pageId + 1 ) )
```
`pageId + 1` should really be `0`, right?

fickludd replied:
That depends on which if condition that got us into this block. Now it is a little bit sneakily depending on `pageId` being -1 when it's not initialized, which I guess should be made more explicit.

chrisvest replied:
Oh, right. Clever, but please add a comment :)

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/standard/NodeRecordFormat.java:
```diff
@@ -68,6 +69,7 @@ public void read( NodeRecord record, PageCursor cursor, RecordLoad mode, int rec
                     BaseRecordFormat.longFromIntAndMod( nextProp, propModifier ), dense,
                     BaseRecordFormat.longFromIntAndMod( nextRel, relModifier ), labels );
         }
+        cursor.setOffset( offset );
```
Is it really necessary to save and restore the offset? The `setOffset` method does bounds checking, which would be nice to avoid if possible.

chrisvest replied:
Also, I notice that you are not modifying any other standard record formats. I don't think this is necessary because `CommonAbstractStore.prepareForReading` already calls `setOffset` in anticipation of the `shouldRetry` loop.

fickludd committed:
Add mark() functionality to PageCursor

The mark allows a page cursor to mark the current offset. The page
cursor offset can later be reset to the mark using setOffsetToMark().
This functionality is useful for retrying reads for example, and has the
added benefit that it setOffsetToMark() does not force a new bounds check.

fickludd committed:
Change CommonAbstractStore to use PageCursor.mark()

fickludd committed:
Improve scan speed by avoiding division and modulus

fickludd committed:
Improve relationship scan speed by using nextRecordByCursor

ragadeeshu commented:
@fickludd Looks like you broke something

fickludd committed:
WIP

fickludd committed:
Minor fixes to make tests happy

fickludd committed:
Adjust broken test

This test would create high-limit records of 38 bytes (exceeding one
physical 32 byte record), without properly preparing the records. This
meant that RelationshipRecordFormat.doReadInternal and doWriteInternal
would take a illegal code path that just happened to work because no
other records were created.

This test is now adjusted to create a record <32 bytes, so it fits
inside one physical record, and now uses the intended code path.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -1051,22 +1054,62 @@ void readIntoRecord( long id, RECORD record, RecordLoad mode, PageCursor cursor
+    {
+        if ( cursor.getOffset() % recordSize != 0 )
+        {
+            System.out.println("hi");
```
hello, how are you?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -1051,22 +1054,62 @@ void readIntoRecord( long id, RECORD record, RecordLoad mode, PageCursor cursor
+        {
+            System.out.println("hi");
+        }
+        assert cursor.getOffset() % recordSize == 0 : "Cursor offset is record start";
```
would be really nice if these guys be not '-ea' enforced checks

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/standard/NodeRecordFormat.java:
```diff
@@ -32,6 +30,7 @@
 {
     // in_use(byte)+next_rel_id(int)+next_prop_id(int)+labels(5)+extra(byte)
     public static final int RECORD_SIZE = 15;
+    private static final int HEADER_SIZE = 1;
```
it looks like this is not a right place for this constant, all of the one byte header records will have header size of 1, not only node records

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/standard/RelationshipRecordFormat.java:
```diff
@@ -35,6 +35,7 @@
     // first_prev_rel_id(int)+first_next_rel_id+second_prev_rel_id(int)+
     // second_next_rel_id+next_prop_id(int)+first-in-chain-markers(1)
     public static final int RECORD_SIZE = 34;
+    private static final int HEADER_SIZE = 1;
```
and here it goes confirmation of above comment

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/newapi/MockStore.java:
```diff
@@ -579,12 +579,24 @@ void node( NodeRecord record, long reference, PageCursor pageCursor )
+    @Override
+    void nodeAdvance( NodeRecord record, PageCursor pageCursor )
+    {
+        initialize( record, record.getId() + 1, nodes );
```
will this work for cases when we do not allocate ids one after another?

fickludd replied:
well if this _MockStore_ suddenly breaks I guess you get to blame me :)

MishaDemianenko commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/highlimit/BaseHighLimitRecordFormat.java:
```diff
@@ -160,6 +160,9 @@ public void read( RECORD record, PageCursor primaryCursor, RecordLoad mode, int
         }
+
+        // Set cursor offset to next record to prepare next read in case of scanning.
+        primaryCursor.setOffset( primaryStartOffset + recordSize );
```
i not really sure why we set cursor offset inside of read here? and not outside, can you please give more wide description of why this needs to be done here and also why in all versions of BaseHighLimitRecordFormat?

fickludd replied:
It's because this optimization relies on every entity record format to leave the cursor offset positioned on the first byte in the next record, after a read. It's done by the read method because this happens anyway during regular reads, so no extra cost is introduces. When std format records are not in use we need to explicitly advance the offset, and for all high-limit formats, because there we never know how much we've read because of the dynamic encoding.

I should probably add a test to ensure this offset behavior... will do that.

chrisvest replied:
See if you can make the test parameterised across formats, and for all stores if possible.

fickludd replied:
Added it to `AbstractRecordFormatTest`, see last commit. That seems to do all current formats for nodes + relationships. Good enough?

chrisvest replied:
Some of those have the checks turned off because of the bounds weirdness with the special cursors we talked about yesterday?

fickludd replied:
Not really, it's just that we only scan nodes and relationships, so all other formats can read as they please. I also haven't modified any other format read code.

MishaDemianenko reviewed (commented):
Sorry, have couple of quick comments

fickludd committed:
Clean up some debugging left-overs

fickludd committed:
Test that node and entity record reads leave the cursor offset correctly

To optimize scan speeds, we now rely on records reads of node and
relationship records to always leave the page cursor offset positioned
on the first byte of the next record.
